shell/platform/linux/fl_keyboard_manager.cc

```diff
@@ -0,0 +1,510 @@
+    snprintf(buffer, kBufferSize, " // 0x%02x", keycode);
+    debug_layout_data.append(buffer);
+  }
+}
+#endif
+
+}  // namespace
+
+/* Define FlKeyboardManagerUserData */
+
+/**
+ * FlKeyboardManagerUserData:
+ * The user_data used when #FlKeyboardManager sends event to
+ * responders.
+ */
+
+struct _FlKeyboardManagerUserData {
+  GObject parent_instance;
+
+  // The owner manager.
+  GWeakRef manager;
+  uint64_t sequence_id;
+};
+
+G_DEFINE_TYPE(FlKeyboardManagerUserData,
+              fl_keyboard_manager_user_data,
+              G_TYPE_OBJECT)
+
+static void fl_keyboard_manager_user_data_dispose(GObject* object) {
+  g_return_if_fail(FL_IS_KEYBOARD_MANAGER_USER_DATA(object));
+  FlKeyboardManagerUserData* self = FL_KEYBOARD_MANAGER_USER_DATA(object);
+
+  g_weak_ref_clear(&self->manager);
+
+  G_OBJECT_CLASS(fl_keyboard_manager_user_data_parent_class)->dispose(object);
+}
+
+static void fl_keyboard_manager_user_data_class_init(
+    FlKeyboardManagerUserDataClass* klass) {
+  G_OBJECT_CLASS(klass)->dispose = fl_keyboard_manager_user_data_dispose;
+}
+
+static void fl_keyboard_manager_user_data_init(
+    FlKeyboardManagerUserData* self) {}
+
+// Creates a new FlKeyboardManagerUserData private class with all information.
+static FlKeyboardManagerUserData* fl_keyboard_manager_user_data_new(
+    FlKeyboardManager* manager,
+    uint64_t sequence_id) {
+  FlKeyboardManagerUserData* self = FL_KEYBOARD_MANAGER_USER_DATA(
+      g_object_new(fl_keyboard_manager_user_data_get_type(), nullptr));
+
+  g_weak_ref_init(&self->manager, manager);
+  self->sequence_id = sequence_id;
+  return self;
+}
+
+/* Define FlKeyboardManager */
+
+struct _FlKeyboardManager {
+  GObject parent_instance;
+
+  GWeakRef view_delegate;
+
+  // An array of #FlKeyResponder. Elements are added with
+  // #fl_keyboard_manager_add_responder immediately after initialization and are
+  // automatically released on dispose.
+  GPtrArray* responder_list;
+
+  // An array of #FlKeyboardPendingEvent.
+  //
+  // Its elements are *not* unreferenced when removed. When FlKeyboardManager is
+  // disposed, this array will be set with a free_func so that the elements are
+  // unreferenced when removed.
+  GPtrArray* pending_responds;
+
+  // An array of #FlKeyboardPendingEvent.
+  //
+  // Its elements are unreferenced when removed.
+  GPtrArray* pending_redispatches;
+
+  // The last sequence ID used. Increased by 1 by every use.
+  uint64_t last_sequence_id;
+
+  // Record the derived layout.
+  //
+  // It is cleared when the platform reports a layout switch. Each entry,
+  // which corresponds to a group, is only initialized on the arrival of the
+  // first event for that group that has a goal keycode.
+  FlKeyboardLayout* derived_layout;
+
+  // A static map from keycodes to all layout goals.
+  //
+  // It is set up when the manager is initialized and is not changed ever after.
+  std::unique_ptr<std::map<uint16_t, const LayoutGoal*>> keycode_to_goals;
```

Comment:
Here and elsewhere, could these be `std::unordered_map`? Unless we need ordering guarantees, generally prefer `std::unordered_map` for performance reasons, not that I expect a lot of this code to be critical to performance.

Reply:
I haven't tackled these data structures yet - they are in the original form. I expect this will move anyway, possibly into a separate class.